### PR TITLE
[SV-COMP'18 5/19] Abort concurrency encoding in possibly unsound cases

### PR DESCRIPTION
**Pull request by tautschnig**

### CODE REVIEW

File: src/goto-symex/goto_symex_state.cpp [hunk 1]
```diff
@@ -342,6 +342,9 @@ void goto_symex_statet::assignment(
   assert_l2_renaming(lhs);
   assert_l2_renaming(rhs);
 
+  if(is_shared && lhs.type().id() == ID_pointer)
```
Comment (smowton) [inline src/goto-symex/goto_symex_state.cpp]: Add a comment here referencing an issue that describes what's wrong, and what we'd need to do to fix it.

Comment (tautschnig) [reply src/goto-symex/goto_symex_state.cpp]: Done.

**Review by smowton — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 26b13ae43fa9fca8078061f987e7087528502142).

File: src/goto-symex/goto_symex_state.cpp [hunk 2]
```diff
@@ -342,6 +342,10 @@ void goto_symex_statet::assignment(
   assert_l2_renaming(lhs);
   assert_l2_renaming(rhs);
 
+  // see #305 on GitHub for a simple example and possible discussion
+  if(is_shared && lhs.type().id() == ID_pointer)
+    throw "pointer handling for concurrency is unsound";
```
Comment (peterschrammel) [inline src/goto-symex/goto_symex_state.cpp:347]: Note: that should be turned into a user facing exception (not an invariant) during the error handling clean up.

**Review by peterschrammel — APPROVED** [review]

**Comment by cesaro** [comment]
@peterschrammel @tautschnig In Java everything is a pointer, which means that pretty much nothing works after this if the program is multithreaded. In particular all tests in the about-to-be-merged https://github.com/diffblue/cbmc/pull/2280 now break. We should have been told about this.

I do support the move to make the user aware that pointer reasoning in a multithread environment is currently unsound. But this means that our work since September 2017 on Java concurrency is unusable after this. The solution is obviously fixing symex, but that will take a while and we have now new priorities.

What about just emitting a warning instead of throwing?

**Comment by tautschnig** [comment]
> We should have been told about this.

This PR has been public for a fair while, so I'm not sure what else should have been said?

Are there reasons that symex is actually sound for Java, despite the limitations? If so, maybe we can tweak the check here to accept the cases where it is sound. The key question for the current implementation being sound is whether a pointer will always be set/updated by a thread with a lower thread id before it is being dereferenced in a thread with the same or a higher thread id.

**Comment by cesaro** [comment]
I cannot imagine any such reason. Java threads can access global variables (static fields) in any possible order, like POSIX threads.

Again, what's the reason for aborting? BMC is in general unsound, and CBMC is probably unsound in some other aspects other than loop bounding. The user is aware of it and can live with it. What's the point of aborting instead of emitting a warning?

**Comment by tautschnig** [comment]
> Again, what's the reason for aborting? BMC is in general unsound, and CBMC is probably unsound in some other aspects other than loop bounding. The user is aware of it and can live with it. What's the point of aborting instead of emitting a warning?

I'm not sure I agree with this: BMC isn't unsound per se if you use loop unwinding assertions, let alone for loop-free programs. The other sources of unsoundness in CBMC, apart from existing-but-not-yet-known bugs, are step-by-step being removed - in fact #1086 is the only other remaining one that I am aware of. So the point is indeed being sound.

Please do push back on the above, I might as well be wrong. Otherwise I'd also be happy to have a command-line option that allows skipping this soundness-check (much like not enabling --unwinding-assertions also takes you into unsound territory). Command-line options enable to user to make a conscious choice.

**Comment by cesaro** [comment]
I do support the move of making the tool as sound as possible, and even making that the default when it's possible. But that has to be balanced with the fact that development is done step by step and people has different priorities. What about not throwing (and instead showing a warning) when `--java-threading` is used. This option was introduced to expose experimental support for concurrency.

Not making this throw a warning (conditional on the presence of a commandline option or by default) makes impossible to continue adding support for concurrency in Java in places other than fixing the very problem that this throw points out.

**Comment by tautschnig** [comment]
I'd happily approve a pull request that adds an option which `--java-threading` sets to true so as to skip this check that got merged here.

**Pull request merged**